extras/minimize_testcase.cpp

```diff
@@ -0,0 +1,344 @@
+      SimplePolygon poly = std::move(polys[i]);
+      polys.erase(polys.begin() + i);
+      try {
+        auto result = Triangulate(polys, precision);
+        polys.insert(polys.begin() + i, std::move(poly));
+      } catch (geometryErr &e) {
+        if (msg.size() > 0 && msg.compare(e.what()) != 0) {
+          polys.insert(polys.begin() + i, std::move(poly));
+        } else {
+          removedSomething = true;
+          msg = e.what();
+        }
+      } catch (topologyErr &e) {
+        polys.insert(polys.begin() + i, std::move(poly));
+      }
+    }
+
+    for (int i = 0; i < polys.size(); i++) {
+      std::vector<int> children = getChildren(polys, i);
+      for (int j = 0; j < polys[i].size(); j++) {
+        // removed vertex cannot change inclusion relation
+        // we just check if the vertex
+        // x: intersects with j0, j (original edge 1)
+        // y: intersects with j, j1 (original edge 2)
+        // z: intersects with j0, j1 (new edge)
+        // if (x ^ y ^ z) is true, it means that the count % 2 is changed,
+        // and we changed inclusion relation, so vertex j cannot be removed
+        int j0 = j == 0 ? polys[i].size() - 1 : j - 1;
+        int j1 = j == polys[i].size() - 1 ? 0 : j + 1;
+        if (std::any_of(children.begin(), children.end(), [&](int k) {
+              return rayHit(polys[k][0], polys[i][j0], polys[i][j]) ^
+                     rayHit(polys[k][0], polys[i][j], polys[i][j1]) ^
+                     rayHit(polys[k][0], polys[i][j0], polys[i][j1]);
+            }))
+          continue;
+
+        // removed vertex cannot introduce intersection
+        if (!safeToRemove(polys, i, j, precision)) continue;
+        glm::vec2 removed = polys[i][j];
+        polys[i].erase(polys[i].begin() + j);
+        try {
+          auto result = Triangulate(polys, precision);
+          polys[i].insert(polys[i].begin() + j, removed);
+        } catch (geometryErr &e) {
+          if (msg.size() > 0 && msg.compare(e.what()) != 0) {
+            polys[i].insert(polys[i].begin() + j, removed);
+          } else {
+            removedSomething = true;
+            msg = e.what();
+          }
+        } catch (topologyErr &e) {
+          polys[i].insert(polys[i].begin() + j, removed);
+        }
+      }
+    }
+  }
+  std::cout << "Error message for triangulation:" << std::endl
+            << msg.c_str() << std::endl;
+}
+
+struct Edge {
+  glm::vec2 west;  // -x side vertex
+  glm::vec2 east;  // +x side vertex
+  size_t i, j;     // indices of origin vertex
+  bool operator<(const Edge &other) const {
+    return west.x < other.west.x ||
+           (west.x == other.west.x && east.x < other.east.x);
+  }
+  Edge(const Polygons &polys, size_t i, size_t j) : i(i), j(j) {
+    size_t j1 = j == polys[i].size() - 1 ? 0 : j + 1;
+    east = polys[i][j];
+    west = polys[i][j1];
+    if (west.x > east.x) std::swap(east, west);
+  }
+};
+
+int isValid(const Polygons &polys, float precision = -1) {
```

Comment:
Do we need this function at all anymore? Or if the goal is to print where the problems are, maybe just list the CW triangles?

Reply:
The goal of this is to check if the input polygon is valid, so we can know if the geometrically invalid error is really correct.